File: cpp/core/src/Vec3.cpp
// Vec3.cpp — the double-precision Vec3 bodies (declared in include/Vec3.hpp).
//
// Do NOT "optimize" these method bodies — parity with the JS oracle depends on
// the exact operation order (float add isn't associative):
//   * normalize() divides by length()||1, so a zero vector stays zero (not NaN).
//   * applyQuaternion() uses r128's inverse-quaternion form (ix/iy/iz/iw), NOT
//     the t = 2*cross form three.js adopted around r150 — they differ in the last
//     ULP.
//   * applyAxisAngle() goes through setFromAxisAngle -> applyQuaternion, as THREE
//     does; it is not a direct Rodrigues rotation.
#include "Vec3.hpp"
#include <cmath>

namespace tox {

Vec3::Vec3(double x_, double y_, double z_) : x(x_), y(y_), z(z_) {}

Vec3& Vec3::set(double x_, double y_, double z_) {
  x = x_;
  y = y_;
  z = z_;
  return *this;
}
Vec3 Vec3::clone() const { return Vec3(x, y, z); }
Vec3& Vec3::copy(const Vec3& v) {
  x = v.x;
  y = v.y;
  z = v.z;
  return *this;
}

Vec3& Vec3::add(const Vec3& v) {
  x += v.x;
  y += v.y;
  z += v.z;
  return *this;
}
Vec3& Vec3::addVectors(const Vec3& a, const Vec3& b) {
  x = a.x + b.x;
  y = a.y + b.y;
  z = a.z + b.z;
  return *this;
}
Vec3& Vec3::addScaledVector(const Vec3& v, double s) {
  x += v.x * s;
  y += v.y * s;
  z += v.z * s;
  return *this;
}
Vec3& Vec3::sub(const Vec3& v) {
  x -= v.x;
  y -= v.y;
  z -= v.z;
  return *this;
}
Vec3& Vec3::subVectors(const Vec3& a, const Vec3& b) {
  x = a.x - b.x;
  y = a.y - b.y;
  z = a.z - b.z;
  return *this;
}
Vec3& Vec3::multiplyScalar(double s) {
  x *= s;
  y *= s;
  z *= s;
  return *this;
}
Vec3& Vec3::divideScalar(double s) { return multiplyScalar(1.0 / s); }

// Rotate about a (unit) axis by angle radians, via a quaternion — THREE's path.
Vec3& Vec3::applyAxisAngle(const Vec3& axis, double angle) {
  const double halfAngle = angle / 2.0, s = std::sin(halfAngle);
  return applyQuaternion(axis.x * s, axis.y * s, axis.z * s, std::cos(halfAngle));
}

// Vector3.applyQuaternion (r128), kept in this exact operation order.
Vec3& Vec3::applyQuaternion(double qx, double qy, double qz, double qw) {
  const double vx = x, vy = y, vz = z;
  // calculate quat * vector
  const double ix = qw * vx + qy * vz - qz * vy;
  const double iy = qw * vy + qz * vx - qx * vz;
  const double iz = qw * vz + qx * vy - qy * vx;
  const double iw = -qx * vx - qy * vy - qz * vz;
  // calculate result * inverse quat
  x = ix * qw + iw * -qx + iy * -qz - iz * -qy;
  y = iy * qw + iw * -qy + iz * -qx - ix * -qz;
  z = iz * qw + iw * -qz + ix * -qy - iy * -qx;
  return *this;
}

double Vec3::dot(const Vec3& v) const { return x * v.x + y * v.y + z * v.z; }
Vec3& Vec3::crossVectors(const Vec3& a, const Vec3& b) {
  const double ax = a.x, ay = a.y, az = a.z;
  const double bx = b.x, by = b.y, bz = b.z;
  x = ay * bz - az * by;
  y = az * bx - ax * bz;
  z = ax * by - ay * bx;
  return *this;
}

double Vec3::lengthSq() const { return x * x + y * y + z * z; }
double Vec3::length() const { return std::sqrt(x * x + y * y + z * z); }

// Zero-length stays zero (divide by 1), never NaN.
Vec3& Vec3::normalize() {
  double l = length();
  return divideScalar(l != 0.0 ? l : 1.0);
}

Vec3& Vec3::lerp(const Vec3& v, double alpha) {
  x += (v.x - x) * alpha;
  y += (v.y - y) * alpha;
  z += (v.z - z) * alpha;
  return *this;
}

double Vec3::distanceToSquared(const Vec3& v) const {
  const double dx = x - v.x, dy = y - v.y, dz = z - v.z;
  return dx * dx + dy * dy + dz * dz;
}
double Vec3::distanceTo(const Vec3& v) const { return std::sqrt(distanceToSquared(v)); }

Vec3& Vec3::negate() {
  x = -x;
  y = -y;
  z = -z;
  return *this;
}

}  // namespace tox

File: js/vec3.js
/* vec3.js — a tiny hand-rolled 3D vector, a byte-for-byte behavioural mirror of
 * THREE.Vector3 *as shipped in three.js r128* (the exact CDN build track.html
 * loads). It exists so the physics core (js/track-physics.js) can be pulled out
 * of the THREE-dependent game and run headless in Node — and so the eventual
 * C++ port has a single, unambiguous vector semantics to transliterate.
 *
 * PARITY IS THE WHOLE POINT. Do NOT "modernize" these method bodies:
 *
 *  - `normalize()` divides by `length() || 1`, so a zero-length vector stays the
 *    zero vector instead of becoming NaN. track-physics relies on this edge case
 *    (see tangentize / beginAirborne).
 *  - `applyQuaternion()` uses r128's inverse-quaternion form (ix/iy/iz/iw), NOT
 *    the `t = 2 * cross(q.xyz, v)` form three.js switched to around r150. The two
 *    are algebraically equal but differ in the last ULP because float add is not
 *    associative — matching r128's *operation order* is what keeps per-step
 *    parity with the shipping game (and, later, with the C++ engine).
 *  - `applyAxisAngle()` goes through setFromAxisAngle → applyQuaternion, exactly
 *    as THREE does; it is NOT a direct Rodrigues rotation.
 *
 * Only the subset of Vector3's surface the track code actually uses is provided.
 * Every method that mutates returns `this`, matching THREE's chaining contract,
 * so call sites transliterate unchanged.
 */

// Module-scratch quaternion, mirroring THREE's internal `_quaternion` reused by
// Vector3.applyAxisAngle so no allocation happens per rotation.
const _q = { x: 0, y: 0, z: 0, w: 1 };

// Quaternion.setFromAxisAngle (r128). `axis` is assumed to be a unit vector.
function setQuatFromAxisAngle(q, axis, angle) {
  const halfAngle = angle / 2, s = Math.sin(halfAngle);
  q.x = axis.x * s;
  q.y = axis.y * s;
  q.z = axis.z * s;
  q.w = Math.cos(halfAngle);
  return q;
}

export class Vec3 {
  constructor(x = 0, y = 0, z = 0) {
    this.x = x;
    this.y = y;
    this.z = z;
  }

  set(x, y, z) {
    this.x = x;
    this.y = y;
    this.z = z;
    return this;
  }

  clone() {
    return new Vec3(this.x, this.y, this.z);
  }

  copy(v) {
    this.x = v.x;
    this.y = v.y;
    this.z = v.z;
    return this;
  }

  add(v) {
    this.x += v.x;
    this.y += v.y;
    this.z += v.z;
    return this;
  }

  addVectors(a, b) {
    this.x = a.x + b.x;
    this.y = a.y + b.y;
    this.z = a.z + b.z;
    return this;
  }

  addScaledVector(v, s) {
    this.x += v.x * s;
    this.y += v.y * s;
    this.z += v.z * s;
    return this;
  }

  sub(v) {
    this.x -= v.x;
    this.y -= v.y;
    this.z -= v.z;
    return this;
  }

  subVectors(a, b) {
    this.x = a.x - b.x;
    this.y = a.y - b.y;
    this.z = a.z - b.z;
    return this;
  }

  multiply(v) {
    this.x *= v.x;
    this.y *= v.y;
    this.z *= v.z;
    return this;
  }

  multiplyScalar(scalar) {
    this.x *= scalar;
    this.y *= scalar;
    this.z *= scalar;
    return this;
  }

  divideScalar(scalar) {
    return this.multiplyScalar(1 / scalar);
  }

  // Rotate about a (unit) axis by `angle` radians, via a quaternion — THREE's
  // exact path, not a direct Rodrigues formula.
  applyAxisAngle(axis, angle) {
    return this.applyQuaternion(setQuatFromAxisAngle(_q, axis, angle));
  }

  // Vector3.applyQuaternion (r128). Kept in this exact operation order on
  // purpose — see the file header.
  applyQuaternion(q) {
    const x = this.x, y = this.y, z = this.z;
    const qx = q.x, qy = q.y, qz = q.z, qw = q.w;

    // calculate quat * vector
    const ix = qw * x + qy * z - qz * y;
    const iy = qw * y + qz * x - qx * z;
    const iz = qw * z + qx * y - qy * x;
    const iw = -qx * x - qy * y - qz * z;

    // calculate result * inverse quat
    this.x = ix * qw + iw * -qx + iy * -qz - iz * -qy;
    this.y = iy * qw + iw * -qy + iz * -qx - ix * -qz;
    this.z = iz * qw + iw * -qz + ix * -qy - iy * -qx;
    return this;
  }

  dot(v) {
    return this.x * v.x + this.y * v.y + this.z * v.z;
  }

  cross(v) {
    return this.crossVectors(this, v);
  }

  crossVectors(a, b) {
    const ax = a.x, ay = a.y, az = a.z;
    const bx = b.x, by = b.y, bz = b.z;
    this.x = ay * bz - az * by;
    this.y = az * bx - ax * bz;
    this.z = ax * by - ay * bx;
    return this;
  }

  lengthSq() {
    return this.x * this.x + this.y * this.y + this.z * this.z;
  }

  length() {
    return Math.sqrt(this.x * this.x + this.y * this.y + this.z * this.z);
  }

  // Zero-length → divideScalar(1): stays the zero vector, never NaN.
  normalize() {
    return this.divideScalar(this.length() || 1);
  }

  setLength(length) {
    return this.normalize().multiplyScalar(length);
  }

  lerp(v, alpha) {
    this.x += (v.x - this.x) * alpha;
    this.y += (v.y - this.y) * alpha;
    this.z += (v.z - this.z) * alpha;
    return this;
  }

  distanceTo(v) {
    return Math.sqrt(this.distanceToSquared(v));
  }

  distanceToSquared(v) {
    const dx = this.x - v.x, dy = this.y - v.y, dz = this.z - v.z;
    return dx * dx + dy * dy + dz * dz;
  }

  negate() {
    this.x = -this.x;
    this.y = -this.y;
    this.z = -this.z;
    return this;
  }

  equals(v) {
    return v.x === this.x && v.y === this.y && v.z === this.z;
  }

  fromArray(array, offset = 0) {
    this.x = array[offset];
    this.y = array[offset + 1];
    this.z = array[offset + 2];
    return this;
  }

  toArray(array = [], offset = 0) {
    array[offset] = this.x;
    array[offset + 1] = this.y;
    array[offset + 2] = this.z;
    return array;
  }
}

export default Vec3;
